storage/store/collections_test.go

```diff
@@ -0,0 +1,139 @@
+		require.NoError(t, err)
+
+		// retrieve the light collection by collection id
+		actual, err := collections.LightByID(expected.ID())
+		require.NoError(t, err)
+
+		// check if the light collection was indeed persisted
+		assert.Equal(t, &expected, actual)
+
+		expectedID := expected.ID()
+
+		// retrieve the collection light id by each of its transaction id
+		for _, txID := range expected.Transactions {
+			collLight, err := collections.LightByTransactionID(txID)
+			actualID := collLight.ID()
+			// check that the collection id can indeed be retrieved by transaction id
+			require.NoError(t, err)
+			assert.Equal(t, expectedID, actualID)
+		}
+
+	})
+}
+
+func TestCollections_IndexDuplicateTx(t *testing.T) {
+	dbtest.RunWithDB(t, func(t *testing.T, db storage.DB) {
+		metrics := metrics.NewNoopCollector()
+		transactions := store.NewTransactions(metrics, db)
+		collections := store.NewCollections(db, transactions)
+
+		// create two collections which share 1 transaction
+		col1 := unittest.CollectionFixture(2)
+		col2 := unittest.CollectionFixture(1)
+		dupTx := col1.Transactions[0]  // the duplicated transaction
+		col2Tx := col2.Transactions[0] // transaction that's only in col2
+		col2.Transactions = append(col2.Transactions, dupTx)
+
+		// insert col1
+		col1Light := col1.Light()
+		err := collections.StoreLightAndIndexByTransaction(&col1Light)
+		require.NoError(t, err)
+
+		// insert col2
+		col2Light := col2.Light()
+		err = collections.StoreLightAndIndexByTransaction(&col2Light)
+		require.NoError(t, err)
+
+		// should be able to retrieve col2 by ID
+		gotLightByCol2ID, err := collections.LightByID(col2.ID())
+		require.NoError(t, err)
+		assert.Equal(t, &col2Light, gotLightByCol2ID)
+
+		// should be able to retrieve col2 by the transaction which only appears in col2
+		_, err = collections.LightByTransactionID(col2Tx.ID())
+		require.NoError(t, err)
+
+		// col2 (not col1) should be indexed by the shared transaction (since col1 was overwritten by col2)
+		gotLightByDupTxID, err := collections.LightByTransactionID(dupTx.ID())
+		require.NoError(t, err)
+		assert.Equal(t, &col1Light, gotLightByDupTxID)
+	})
+}
+
+// verify that when StoreLightAndIndexByTransaction is concurrently called with same tx and
+// different collection both will succeed, and one of the collection will be indexed by the tx
+func TestCollections_ConcurrentIndexByTx(t *testing.T) {
+	dbtest.RunWithDB(t, func(t *testing.T, db storage.DB) {
+		metrics := metrics.NewNoopCollector()
+		transactions := store.NewTransactions(metrics, db)
+		collections := store.NewCollections(db, transactions)
+
+		// Create two collections sharing the same transaction
+		col1 := unittest.CollectionFixture(1)
+		col2 := unittest.CollectionFixture(1)
+		sharedTx := col1.Transactions[0] // The shared transaction
+		col2.Transactions[0] = sharedTx
+
+		var wg sync.WaitGroup
+		wg.Add(2)
+
+		errChan := make(chan error, 2)
+
+		// Concurrently insert col1
+		go func() {
```

Comment:
this feels like it might pass by chance even if there is an issue. Can we have 2 go-routines each inserting 100 collections?